File: stack/remove_duplicate_letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        last_index = {}

        for i in range(len(s)):
            last_index[s[i]] = i

        stack = []
        visited = set()

        for i in range(len(s)):
            if s[i] not in visited:
                # monotonic increasing stack
                while stack and stack[-1] > s[i] and last_index[stack[-1]] > i:
                    visited.remove(stack.pop())
                stack.append(s[i])
                visited.add(s[i])

        return "".join(stack)

File: stack/remove_duplicate_letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        HashMap<Character, Integer> last_index = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            last_index.put(s.charAt(i), i);
        }

        Stack<Character> stack = new Stack<>();
        HashSet<Character> visited = new HashSet<>();

        for (int i = 0; i < s.length(); i++) {
            char currentChar = s.charAt(i);
            if (!visited.contains(currentChar)) {
                while (!stack.isEmpty() && stack.peek() > currentChar && last_index.get(stack.peek()) > i) {
                    visited.remove(stack.pop());
                }
                stack.push(currentChar);
                visited.add(currentChar);
            }
        }

        StringBuilder result = new StringBuilder();
        for (char c : stack) {
            result.append(c);
        }
        return result.toString();

    }
}